llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp

```diff
@@ -0,0 +1,244 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLPropagateJointMatrixUsage.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/IntrinsicInst.h"
+
+#include <set>
+
+using namespace llvm;
+
+namespace {
+
+/// Returns true if the function is a SPIRV or SYCL builtin, e.g.
+/// _Z28__spirv_GlobalInvocationId_xv
+/// NB! This function was copied from sycl-post-link/ModuleSplitter.cpp and the
+/// definition of entry point (i.e. implementation of the function) should be in
+/// sync between those two.
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.startswith("__spirv_") || FName.startswith("__sycl_");
+}
+
+bool isEntryPoint(const Function &F) {
+  // Skip declarations, we can't analyze them
+  if (F.isDeclaration()) {
+    // F.print(outs());
+    return false;
+  }
+
+  // Kernels are always considered to be entry points
+  if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+    return true;
+
+  // FIXME: sycl-post-link allows to disable treating SYCL_EXTERNAL's as entry
+  // points - do we need similar flag here?
+  // SYCL_EXTERNAL functions with sycl-module-id attribute
+  // are also considered as entry points (except __spirv_* and __sycl_*
+  // functions)
+  return F.hasFnAttribute("sycl-module-id") && !isSpirvSyclBuiltin(F.getName());
+}
+
+using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
+
+/// Updates call graph with the information from function @F
+void fillCallGraph(Function *F, CallGraphTy &CG) {
+  for (Instruction &I : instructions(F)) {
+    if (const auto *CI = dyn_cast<CallInst>(&I)) {
+      if (!CI->isIndirectCall() && CI->getCalledFunction())
+        CG[F].insert(CI->getCalledFunction());
+    }
+  }
+}
+
+using JointMatrixValuesSetTy = std::set<SmallString<24>>;
+using FunctionToJointMatrixValuesMapTy =
+    DenseMap<Function *, JointMatrixValuesSetTy>;
+
+/// Creates mapping between a function and an information about matrix types and
+/// sizes of sycl::ext::oneapi::experimental::matrix::joint_matrix type
+void fillFunctionToJointMatrixValuesMap(
+    Function *F,
+    FunctionToJointMatrixValuesMapTy &FunctionToJointMatrixValues) {
+  // assume we have other sycl-joint-matrix-* attributes if
+  // sycl-joint-matrix-type is present
+  if (!F->hasFnAttribute("sycl-joint-matrix-type"))
+    return;
+
+  SmallString<24> Result;
+  // NB! The order of attributes must not change as it is used later in SYCL
+  // RT
+  // The order is:
+  //   - sycl-joint-matrix-type
+  //   - sycl-joint-matrix-use
+  //   - sycl-joint-matrix-rows
+  //   - sycl-joint-matrix-cols
+  // NB! Values must be separated with a comma
+  Result += F->getFnAttribute("sycl-joint-matrix-type").getValueAsString();
```

Comment:
should we add scope as well here? so if a matrix is declared as 
joint_matrix<work_group, half, use::a, 8,16, layout::row_major> A;
this is not supported as only sub_group Group is supported.

Reply:
I agree, I will track adding this in the spec. We removed scope from the spec as we only have one value supported.

Reply:
That specification is not for error checking, it is for understanding which features are used by a certain kernel. 

Checking that `work_group` is a wrong value to pass into the first template arguments could be done with a simple `static_assert` in headers, no need to bring this into the compiler.